test/e2e/network/service.go

```diff
@@ -3283,42 +3282,72 @@ var _ = SIGDescribe("ESIPP [Slow]", func() {
 			framework.Failf("Service HealthCheck NodePort still present")
 		}
 
-		endpointNodeMap, err := jig.GetEndpointNodes()
+		epNodes, err := jig.ListNodesWithEndpoint()
 		framework.ExpectNoError(err)
-		noEndpointNodeMap := map[string][]string{}
+		// map from name of nodes with endpoint to internal ip
+		// it is assumed that there is only a single node with the endpoint
+		endpointNodeMap := make(map[string]string)
+		// map from name of nodes without endpoint to internal ip
+		noEndpointNodeMap := make(map[string]string)
+		for _, node := range epNodes {
+			ips := e2enode.GetAddresses(&node, v1.NodeInternalIP)
+			if len(ips) < 1 {
+				framework.Failf("No internal ip found for node %s", node.Name)
+			}
+			endpointNodeMap[node.Name] = ips[0]
+		}
 		for _, n := range nodes.Items {
-			if _, ok := endpointNodeMap[n.Name]; ok {
-				continue
+			ips := e2enode.GetAddresses(&n, v1.NodeInternalIP)
+			if len(ips) < 1 {
+				framework.Failf("No internal ip found for node %s", n.Name)
+			}
+			if _, ok := endpointNodeMap[n.Name]; !ok {
+				noEndpointNodeMap[n.Name] = ips[0]
 			}
-			noEndpointNodeMap[n.Name] = e2enode.GetAddresses(&n, v1.NodeExternalIP)
 		}
+		framework.ExpectNotEqual(len(endpointNodeMap), 0)
+		framework.ExpectNotEqual(len(noEndpointNodeMap), 0)
 
 		svcTCPPort := int(svc.Spec.Ports[0].Port)
 		svcNodePort := int(svc.Spec.Ports[0].NodePort)
 		ingressIP := e2eservice.GetIngressPoint(&svc.Status.LoadBalancer.Ingress[0])
 		path := "/clientip"
+		dialCmd := "clientip"
 
-		ginkgo.By(fmt.Sprintf("endpoints present on nodes %v, absent on nodes %v", endpointNodeMap, noEndpointNodeMap))
-		for nodeName, nodeIPs := range noEndpointNodeMap {
-			ginkgo.By(fmt.Sprintf("Checking %v (%v:%v%v) proxies to endpoints on another node", nodeName, nodeIPs[0], svcNodePort, path))
-			GetHTTPContent(nodeIPs[0], svcNodePort, e2eservice.KubeProxyLagTimeout, path)
-		}
+		config := e2enetwork.NewNetworkingTestConfig(f, false, false)
```

Comment:
And second thing: as part of NewNetworkingTestConfig above, what happens we also require external IP address:
https://github.com/kubernetes/kubernetes/blob/72fb952dd6cf54f56925853005198957aaf1584d/test/e2e/framework/network/utils.go#L649

@mm4tt - I think this will work, because we have public heapster node, right? Or should it also be changed?

Reply:
I am not pretty sure, but from these lines https://github.com/kubernetes/kubernetes/blob/72fb952dd6cf54f56925853005198957aaf1584d/test/e2e/framework/network/utils.go#L671-L674, it looks like `config.ExternalAddr` is only used to determine `config.NodeIP` and `config.NodeIP` would be set to an internal IP if `config.ExternalAddr` is empty.

Reply:
Yes, we have a heapster node with public ip, and will always have it in our tests (e.g. to provide ssh bastion functionality)